src/Protocol/ForgeHandshake.cpp

```diff
@@ -0,0 +1,365 @@
+				ClientModsString.append(item.second);
+				ClientModsString.append(", ");
+			}
+
+			LOG("Client connected with %zu mods: %s", ClientMods.size(), ClientModsString.c_str());
+
+			if (m_Client->m_ForgeMods != nullptr)
+			{
+				m_Client->m_ForgeMods.reset();
+			}
+
+			m_Client->m_ForgeMods = cpp14::make_unique<cForgeMods>(ClientMods);
+
+			// Let the plugins know about this event, they may refuse the player:
+			if (cRoot::Get()->GetPluginManager()->CallHookLoginForge(*a_Client))
+			{
+				LOG("Modded client refused by plugin");
+				SetError();
+				return;
+			}
+
+			// Send server ModList
+
+			// TODO: max size?
+			cByteBuffer Buf(1024);
+
+			// Send server-side Forge mods registered by plugins
+			auto &ServerMods = m_Client->GetForgeMods();
+
+			size_t ModCount = ServerMods.GetNumMods();
+
+			Buf.WriteBEInt8(Discriminator_ModList);
+			Buf.WriteVarInt32(static_cast<UInt32>(ModCount));
+			for (size_t i = 0; i < ModCount; ++i)
+			{
+				const AString & name = ServerMods.GetModNameAt(i);
+				const AString & version = ServerMods.GetModVersion(name);
+				Buf.WriteVarUTF8String(name);
+				Buf.WriteVarUTF8String(version);
+			}
+			AString ServerModList;
+			Buf.ReadAll(ServerModList);
+
+			m_Client->SendPluginMessage("FML|HS", ServerModList);
+			break;
+		}
+
+		case Discriminator_HandshakeAck:
+		{
+			if (a_Size != 2)
+			{
+				LOG("Unexpected HandshakeAck packet length: %zu", a_Size);
+				SetError();
+				break;
+			}
+
+			int Phase = a_Data[1];
+			LOGD("Received client HandshakeAck with phase=%d", Phase);
+
+			switch (Phase)
+			{
+				case ClientPhase_WAITINGSERVERDATA:
+				{
+					cByteBuffer Buf(1024);
+					Buf.WriteBEInt8(Discriminator_RegistryData);
+
+					// TODO: send real registry data
+					bool HasMore = false;
+					AString RegistryName = "potions";
+					UInt32 NumIDs = 0;
+					UInt32 NumSubstitutions = 0;
+					UInt32 NumDummies = 0;
+
+					Buf.WriteBool(HasMore);
+					Buf.WriteVarUTF8String(RegistryName);
+					Buf.WriteVarInt32(NumIDs);
+					Buf.WriteVarInt32(NumSubstitutions);
+					Buf.WriteVarInt32(NumDummies);
+
+					AString RegistryData;
+					Buf.ReadAll(RegistryData);
+					m_Client->SendPluginMessage("FML|HS", RegistryData);
+					break;
+				}
+
+				case ClientPhase_WAITINGSERVERCOMPLETE:
+				{
+					LOG("Client finished receiving registry data; acknowledging");
+
+					AString Ack;
+					Ack.push_back(Discriminator_HandshakeAck);
+					Ack.push_back(ServerPhase_WAITINGCACK);
+					m_Client->SendPluginMessage("FML|HS", Ack);
+					break;
+				}
+
+				case ClientPhase_PENDINGCOMPLETE:
+				{
+					LOG("Client is pending completion; sending complete ack");
+
+					AString Ack;
+					Ack.push_back(Discriminator_HandshakeAck);
+					Ack.push_back(ServerPhase_COMPLETE);
+					m_Client->SendPluginMessage("FML|HS", Ack);
+
+					// Now finish logging in
+					m_Client->FinishAuthenticate(*m_Name, *m_UUID, *m_Properties);
+					break;
+				}
+			}
+			break;
+		}
```

Comment:
This `switch` makes the function rather large and contains lots of rather unrelated stuff. Especially with the nested `switch`.

I'd move each `case`'s contents into a separate function, then compress the `switch` to single-line `case`s:
```cpp
switch (a_Data[0])
{
  case Discriminator::ClientHello: return HandleClientHello(a_Data, a_Size);
  ...
}
```